File: rec-core/src/main/java/net/kimleo/rec/rule/RuleRepository.java
package net.kimleo.rec.rule;

import net.kimleo.rec.rule.impl.Exist;
import net.kimleo.rec.rule.impl.Unique;

import java.util.HashMap;
import java.util.Map;

public class RuleRepository {
    private final Map<String, RecRule> rules;

    public RuleRepository(Map<String, RecRule> rules) {
        this.rules = rules;
    }

    public RuleRepository() {
        this(new HashMap<>());
        register("unique", new Unique());
        register("exist", new Exist());
    }

    public void register(String name, RecRule rule) {
        rules.put(name, rule);
    }

    public RecRule rule(String name) {
        return rules.get(name);
    }

}

File: src/main/kotlin/net/kimleo/rec/rule/RuleRepository.kt
package net.kimleo.rec.rule

import net.kimleo.rec.orElse
import net.kimleo.rec.rule.impl.Exist
import net.kimleo.rec.rule.impl.Unique
import java.util.*

class RuleRepository(val rules: HashMap<String, RecRule> = hashMapOf()) {

    init {
        register("unique", Unique())
        register("exist", Exist())
    }

    fun register(name: String, rule: RecRule) {
        rules[name] = rule
    }

    fun rule(name: String): RecRule {
        return rules[name].orElse { throw RuntimeException("Cannot found rule $name") }
    }
}